core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+    /// Retrieve the gRPC service name and method name from a URI.
+    ///
+    /// * `uri` - The uri used for the gRPC call.
+    fn retrieve_grpc_names_from_uri(uri: &Uri) -> (String, String) {
+        let uri_string = uri.to_string();
+        let uri_parts: Vec<&str> = uri_string.split('/').collect();
+        let mut service_name = String::new();
+        let mut method_name = String::new();
+        if uri_parts.len() == 5 {
+            method_name = uri_parts[4].to_string();
+            let qualified_service_name = uri_parts[3].to_string();
+            let qualified_service_name_parts: Vec<&str> =
+                qualified_service_name.split('.').collect();
+            if qualified_service_name_parts.len() == 2 {
+                service_name = qualified_service_name_parts[1].to_string();
+            }
+        }
+        (service_name, method_name)
+    }
+}
+
+impl<S> Service<http::request::Request<tonic::transport::Body>> for GrpcInterceptorService<S>
+where
+    S: Service<
+            http::request::Request<tonic::transport::Body>,
+            Response = http::response::Response<tonic::body::BoxBody>,
+            Error = Box<dyn std::error::Error + Sync + Send>,
+        > + Send,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    /// Implementation of tower's Service trait's poll_ready method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    /// Implementation of tower's Service trait's call method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn call(
+        &mut self,
+        mut request: http::request::Request<tonic::transport::Body>,
+    ) -> Self::Future {
+        let interceptor = (self.interceptor_factory)();
+
+        let (service_name, method_name) = Self::retrieve_grpc_names_from_uri(request.uri());
+        let is_applicable = interceptor.is_applicable(&service_name, &method_name);
+
+        if is_applicable && interceptor.must_handle_request() {
+            let (parts, body) = request.into_parts();
+            let mut body_bytes: Bytes =
+                match futures::executor::block_on(hyper::body::to_bytes(body)) {
+                    Ok(bytes) => bytes,
+                    Err(err) => {
+                        return Box::pin(async move {
+                            Err(Box::new(err) as Box<dyn std::error::Error + Sync + Send>)
+                        })
+                    }
+                };
+            let protobuf_message_bytes: Bytes = body_bytes.split_off(GRPC_HEADER_LENGTH);
+            let grpc_header_bytes = body_bytes;
+            let new_protobuf_message_bytes: Bytes = match interceptor.handle_request(
```

Comment:
similar to another comment, this can be done with `map_err` and `?`:

```rust
let new_protobuf_message_bytes = interceptor.handle_request(...)
    .map_err(|err| Box::pin(async move { Err(err) }))?;
```

Reply:
see my other comment, I don't think this will work